url29ecd5d728_geodb_db3dcore_tgz-pJ8-de_uos_igf_db3d_dbms_model3d_SegmentNetBuilderJ8/src/de/uos/igf/db3d/dbms/util/TriangleServices.java

```diff
@@ -0,0 +1,414 @@
+		TriangleElt3D triangle;
+		Point3D[] p;
+		int id = 0;
+
+		// save the IDs of Points to create the Triangles:
+		Set<Point3D> unique = new HashSet<Point3D>();
+
+		for (TriangleNet3DComp tri : comp) {
+
+			Set s = tri.getElementsViaSAM();
+			Iterator<Equivalentable> it = s.iterator();
+			int idForPoint = 0;
+
+			while (it.hasNext()) {
+
+				triangle = (TriangleElt3D) it.next();
+
+				for (Segment3D seg : triangle.getSegments()) {
+					segments.add(seg);
+				}
+
+				p = triangle.getPoints();
+
+				int[] pointsForTriangles = new int[3];
+
+				for (int i = 0; i < 3; i++) {
+
+					if (!unique.contains(p[i])) {
+						unique.add(p[i]);
+						points.put(id, p[i]);
+						pointIDs.put(p[i], id);
+						pointsForTriangles[i] = id;
+
+						pointToTriangle.put(id, new LinkedList<Triangle3D>());
+						pointToTriangle.get(id).add(triangle);
+
+						id++;
+					} else {
+						idForPoint = pointIDs.get(p[i]);
+						pointsForTriangles[i] = idForPoint;
+						pointToTriangle.get(idForPoint).add(triangle);
+					}
+				}
+				triangles.put(triangle.getID(), pointsForTriangles);
+				components.put(triangle.getID(), tri.getID());
+			}
+		}
+	}
+
+	/**
+	 * This methods creates unique points for a 4D triangle net
+	 * 
+	 * @param triangle4dNet
+	 *            - Map with IDs and Triangle4D objects (just point IDs)
+	 * @param pointTubes4D
+	 *            - Point3D objects with an ID
+	 */
+	public void initFor4DPointCloud(Map<Integer, Element4D> triangle4dNet,
+			Map<Integer, Point3D> pointTubes4D) {
+
+		// save the IDs of Points to create the Triangles:
+		Set<Point3D> unique = new HashSet<Point3D>();
+
+		for (Integer triangleID : triangle4dNet.keySet()) {
+
+			Triangle4D tmpTriangle = (Triangle4D) triangle4dNet.get(triangleID);
+
+			int point4DIDs[] = { tmpTriangle.getIDzero(),
+					tmpTriangle.getIDone(), tmpTriangle.getIDtwo() };
+
+			Point3D zero = pointTubes4D.get(point4DIDs[0]);
+			Point3D one = pointTubes4D.get(point4DIDs[1]);
+			Point3D two = pointTubes4D.get(point4DIDs[2]);
+
+			Point3D[] p = { zero, one, two };
+
+			int[] pointsForTriangles = new int[3];
+
+			for (int i = 0; i < 3; i++) {
+
+				if (!unique.contains(p[i])) {
+					unique.add(p[i]);
+					points.put(point4DIDs[i], p[i]);
+					pointIDs.put(p[i], point4DIDs[i]);
+					pointsForTriangles[i] = point4DIDs[i];
+				} else {
+					pointsForTriangles[i] = pointIDs.get(p[i]);
+				}
+			}
+			triangles.put(triangleID, pointsForTriangles);
+		}
+	}
+
+	public void createBoundaryElements(TriangleNet4D net,
+			Map<Integer, Point3D> pointTubes4D, Date date) {
+
+		// save the unique Segments of this TriangleNet
+		Map<Segment4D, Integer> uniqueSegments = new HashMap<Segment4D, Integer>();
+
+		// first segID
+		int segID = net.getBoundaryElements1D(date).size();
+
+		Map<Integer, Element4D> triangleElements = net.getNetElements(date);
+
+		for (Integer triangleID : triangleElements.keySet()) {
+
+			Triangle4D tmpTriangle = (Triangle4D) triangleElements
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.